File: python/examples/core/08-cross-canvas-clone/main.py
"""Example 08: clone a widget from one canvas to another.

Demonstrates the SDK's :meth:`Client.widgets.clone` helper, which implements
the contract added in API changelog §1: cross-canvas cloning is a behaviour
of the *per-type* create endpoints, not a separate ``/widgets/clone`` URL.

Steps:
    1. Fetch the source widget to discover its ``widget_type``.
    2. Call ``client.widgets.clone(dest_canvas_id, src_canvas_id,
       src_widget_id, widget_type)``.
    3. Print the new widget id on the destination canvas.
    4. If ``CANVUS_CLEANUP=1`` is set, pause 2 s then delete the clone.
"""

from __future__ import annotations

import asyncio
import os
import sys

import structlog
from pydantic import Field, ValidationError
from pydantic_settings import BaseSettings, SettingsConfigDict

from canvus_sdk import APIError, Client, configure_logging

logger = structlog.get_logger(__name__)

# Mapping from server-supplied widget_type to the per-type DELETE path used
# for cleanup. Matches the same set the SDK's clone helper supports.
WIDGET_TYPE_TO_PATH: dict[str, str] = {
    "Note": "notes",
    "Image": "images",
    "Video": "videos",
    "Pdf": "pdfs",
    "Browser": "browsers",
    "Anchor": "anchors",
    "Table": "tables",
}


class CloneSettings(BaseSettings):
    """Environment configuration for the cross-canvas clone example."""

    model_config = SettingsConfigDict(
        env_prefix="CANVUS_",
        env_file=".env",
        env_file_encoding="utf-8",
        extra="ignore",
    )

    api_url: str = Field(...)
    api_key: str = Field(...)
    canvas_id: str = Field(..., description="Source canvas id.")
    dest_canvas_id: str = Field(..., description="Destination canvas id.")
    source_widget_id: str = Field(..., description="Widget to clone from source.")


async def _delete_clone(
    client: Client,
    dest_canvas_id: str,
    new_widget_id: str,
    widget_type: str,
) -> None:
    """Best-effort cleanup of the cloned widget."""
    path = WIDGET_TYPE_TO_PATH.get(widget_type)
    if path is None:
        logger.warning(
            "cleanup skipped — unknown widget_type",
            widget_type=widget_type,
        )
        return
    # Reach into the appropriate per-type resource by attribute name. The
    # SDK exposes them on `client.widgets` (notes, images, videos, …).
    resource = getattr(client.widgets, path, None)
    if resource is None or not hasattr(resource, "delete"):
        logger.warning("cleanup skipped — no resource for path", path=path)
        return
    try:
        await resource.delete(dest_canvas_id, new_widget_id)
        logger.info("cleanup deleted", new_widget_id=new_widget_id)
    except APIError as e:
        logger.warning(
            "cleanup delete failed",
            status_code=e.status_code,
            new_widget_id=new_widget_id,
        )


async def main() -> int:
    """Entry point."""
    configure_logging()
    try:
        settings = CloneSettings()  # type: ignore[call-arg]
    except ValidationError as e:
        logger.error("config invalid", errors=e.errors())
        return 2

    cleanup = os.environ.get("CANVUS_CLEANUP") == "1"

    async with Client(settings.api_url, settings.api_key) as client:
        # Step 1: fetch source widget to discover its type.
        try:
            src = await client.widgets.get(settings.canvas_id, settings.source_widget_id)
        except APIError as e:
            logger.error(
                "fetch source failed",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1
        widget_type = src.widget_type or ""
        if widget_type not in WIDGET_TYPE_TO_PATH:
            logger.error(
                "source widget type is not cloneable",
                widget_type=widget_type,
                allowed=sorted(WIDGET_TYPE_TO_PATH.keys()),
            )
            return 1
        logger.info(
            "source fetched",
            widget_id=src.id,
            widget_type=widget_type,
        )

        # Step 2: clone via the SDK helper.
        try:
            cloned = await client.widgets.clone(
                settings.dest_canvas_id,
                settings.canvas_id,
                settings.source_widget_id,
                widget_type,
            )
        except APIError as e:
            logger.error(
                "clone failed",
                status_code=e.status_code,
                response_body=e.response_body,
            )
            return 1
        new_id = cloned.get("id") if isinstance(cloned, dict) else None
        logger.info(
            "cloned",
            source_canvas_id=settings.canvas_id,
            dest_canvas_id=settings.dest_canvas_id,
            new_widget_id=new_id,
        )
        print(
            f"cloned widget {settings.source_widget_id} ({widget_type}) "
            f"into canvas {settings.dest_canvas_id} as {new_id}",
        )

        # Step 4: optional cleanup
        if cleanup and isinstance(new_id, str):
            await asyncio.sleep(2.0)
            await _delete_clone(client, settings.dest_canvas_id, new_id, widget_type)

    return 0


if __name__ == "__main__":
    sys.exit(asyncio.run(main()))

File: go/examples/core/08-cross-canvas-clone/main.go
// Command 08-cross-canvas-clone clones a single widget from one Canvus
// canvas to another using the SDK's CloneWidget helper, which exercises
// the per-type create endpoint with {source_canvas_id, source_widget_id}
// fields per the Phase 3 changelog §1 documentation update.
//
// The legacy POST /api/v1/canvases/{id}/widgets/clone endpoint returns
// 501 Not Implemented and is deliberately not used.
package main

import (
	"context"
	"fmt"
	"log/slog"
	"os"
	"strings"
	"time"

	"github.com/jaypaulb/MT-Canvus-Tools/go/sdk/canvus"
)

// widgetTypeToPath maps the lowercase widget_type discriminator returned by
// the server to the URL path segment used by the per-type create endpoint.
// Asset-based and primitive widget types share the same shape: type-name
// plural-ised.
var widgetTypeToPath = map[string]string{
	"note":    "notes",
	"image":   "images",
	"video":   "videos",
	"pdf":     "pdfs",
	"browser": "browsers",
	"anchor":  "anchors",
	"table":   "tables",
}

func main() {
	setupLogging()
	if err := run(); err != nil {
		slog.Error("cross-canvas-clone failed", "err", err)
		os.Exit(1)
	}
}

func run() error {
	baseURL, err := mustEnv("CANVUS_BASE_URL")
	if err != nil {
		return err
	}
	apiKey, err := mustEnv("CANVUS_API_KEY")
	if err != nil {
		return err
	}
	srcCanvasID, err := mustEnv("CANVUS_CANVAS_ID")
	if err != nil {
		return err
	}
	dstCanvasID, err := mustEnv("CANVUS_DEST_CANVAS_ID")
	if err != nil {
		return err
	}
	srcWidgetID, err := mustEnv("CANVUS_SOURCE_WIDGET_ID")
	if err != nil {
		return err
	}

	cfg := &canvus.SessionConfig{BaseURL: baseURL}
	s := canvus.NewSession(cfg, canvus.WithAPIKey(apiKey))
	ctx := context.Background()

	// Step 1 — fetch the source widget to learn its type.
	src, err := s.GetWidget(ctx, srcCanvasID, srcWidgetID)
	if err != nil {
		return fmt.Errorf("GetWidget (source): %w", err)
	}
	slog.Info("source widget fetched",
		"widget_id", src.ID,
		"widget_type", src.WidgetType,
		"location", pointString(src.Location))

	// Step 2 — resolve widget type → URL path segment.
	wtype := strings.ToLower(src.WidgetType)
	pathSeg, ok := widgetTypeToPath[wtype]
	if !ok {
		return fmt.Errorf("widget type %q is not cloneable via this example "+
			"(supported: note/image/video/pdf/browser/anchor/table)", src.WidgetType)
	}

	// Step 3 — call the SDK clone helper. Passing nil for location keeps
	// the source widget's original coordinates on the destination.
	cloned, err := s.CloneWidget(ctx, dstCanvasID, srcCanvasID, srcWidgetID, pathSeg, nil)
	if err != nil {
		return fmt.Errorf("CloneWidget: %w", err)
	}
	slog.Info("widget cloned",
		"source_widget_id", srcWidgetID,
		"cloned_widget_id", cloned.ID,
		"dest_canvas_id", dstCanvasID,
		"widget_type", cloned.WidgetType,
		"location", pointString(cloned.Location))

	fmt.Printf("cloned widget id on destination: %s\n", cloned.ID)

	// Step 4 — optional cleanup.
	if os.Getenv("CANVUS_CLEANUP") == "1" {
		slog.Info("CANVUS_CLEANUP=1 — sleeping 2s then deleting cloned widget")
		time.Sleep(2 * time.Second)
		if err := s.DeleteWidget(ctx, dstCanvasID, cloned.ID, wtype); err != nil {
			return fmt.Errorf("DeleteWidget (cleanup): %w", err)
		}
		slog.Info("cloned widget deleted", "cloned_widget_id", cloned.ID)
	} else {
		slog.Info("CANVUS_CLEANUP unset — leaving cloned widget on destination canvas")
	}
	return nil
}

func pointString(p *canvus.Point) string {
	if p == nil {
		return "<nil>"
	}
	return fmt.Sprintf("(%.1f, %.1f)", p.X, p.Y)
}

func mustEnv(name string) (string, error) {
	v := os.Getenv(name)
	if v == "" {
		return "", fmt.Errorf("missing required env var %s", name)
	}
	return v, nil
}

func setupLogging() {
	var h slog.Handler
	switch os.Getenv("LOG_FORMAT") {
	case "json":
		h = slog.NewJSONHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	default:
		h = slog.NewTextHandler(os.Stderr, &slog.HandlerOptions{Level: slog.LevelInfo})
	}
	slog.SetDefault(slog.New(h))
}
